### DISCUSSION TITLE
Ignore api and middleware when nextConfig.output is "export"

### DISCUSSION THREAD

**Discussion by zecka** [vercel/next.js#51036] [Help]
### Summary

I have a next13 project that need different build type depending on environnement variable.

I'm using "app" folder

Here is my next.config.js
```js
const nextConfig = {
    output: process.env.NEXT_OUTPUT_MODE,
}
module.exports = nextConfig
```
I have api folder and middleware file that should be used when next is build with output "standalone"

Druring build on "export" mode I got following error

> Export encountered errors on following paths:
        /api/revalidate/route: /api/revalidate

## Reproducible repository
https://github.com/zeckaissue/next-export-api-crash/

**Comment by sahariar-safin** [comment]
Sure, you can use the `exclude` property in your `next.config.js` file to ignore the `api` folder and `middleware` file when the output mode is `export`. Here is an example:

```js
const nextConfig = {
  output: process.env.NEXT_OUTPUT_MODE,
  exclude: [
    'api',
    'middleware',
  ],
}

module.exports = nextConfig
```

This will tell Next.js to ignore the `api` and `middleware` folders when the output mode is `export`. This will prevent the error that you are getting.

Here is a more detailed explanation of how the `exclude` property works. When Next.js builds your application, it recursively walks through the directory structure of your application and looks for files that match the patterns in the `exclude` property. If a file matches a pattern in the `exclude` property, it will be ignored by Next.js.

In your case, you want to ignore the `api` and `middleware` folders. These folders are not important when the output mode is `export`. So, you can add them to the `exclude` property to tell Next.js to ignore them.

I hope this helps!

**Reply by zecka** [reply]
thank's @shafinthedeveloper 

But, the exclude property seems to not be taken into account.

When I add the exclude property to nextConfig, I get the following warning:

- warn     The root value has an unexpected property, exclude, which is not in the list of allowed properties (amp, analyticsId, assetPrefix, basePath, cleanDistDir, compiler, compress, configOrigin, crossOrigin, devIndicators, distDir, env, eslint, excludeDefaultMomentLocales, experimental, exportPathMap, generateBuildId, generateEtags, headers, httpAgentOptions, i18n, images, modularizeImports, onDemandEntries, optimizeFonts, output, outputFileTracing, pageExtensions, poweredByHeader, productionBrowserSourceMaps, publicRuntimeConfig, reactStrictMode, redirects, rewrites, sassOptions, serverRuntimeConfig, skipMiddlewareUrlNormalize, skipTrailingSlashRedirect, staticPageGenerationTimeout, swcMinify, target, trailingSlash, transpilePackages, typescript, useFileSystemPublicRoutes, webpack)

**Reply by jeffreyschultz** [reply]
`exclude` doesn't seem to be a valid `next.config.js` option.

**Reply by knoefel** [reply]
I'm having the same problem when i try to add the `exclude` option. @zecka @jeffreyschultz Did you solve this issue?

**Reply by icyJoseph** [reply]
You can solve it like this I think, with custom extensions: https://github.com/vercel/next.js/discussions/51891#discussioncomment-6297178

**Reply by TeemuKoivisto** [reply]
It seems there's no option to flag this as AI-generated but clearly somebody copy-pasted the question to ChatGPT which produced this nonsense answer of `exclude` property

**Comment by zecka** [comment]
I find a workaround but maybe a built-in solution exist.

I add have installed ignore-loader for webpack

```
yarn add -D ignore-loader
```

Then i add following webpack property to my next.js.config

```ts
const nextConfig = {
  output: process.env.NEXT_OUTPUT_MODE,
  webpack: config => {
    if (process.env.NEXT_OUTPUT_MODE !== 'export' || !config.module) {
      return config;
    }
    config.module.rules?.push({
      test: /src\/app\/api/,
      loader: 'ignore-loader',
    });
    return config;
  },
}
```

**Reply by vietanhdev** [reply]
This does not work for me on Nextjs 14:
```
webpack: config => {
    if (!process.env.EXPORT_STATIC || !config.module) {
      return config;
    }
    config.module.rules?.push({
      test: /app\/api/,
      loader: 'ignore-loader',
    });
    return config;
  },
```
```
Error: Page "/api/auth/[...nextauth]" is missing "generateStaticParams()" so it cannot be used with "output: export" config.
`

**Reply by fatalegayatri** [reply]
do you find any solution

**Comment by pix4bom10** [comment]
> This does not work for me on Nextjs 14:
> 
> ```
> webpack: config => {
>     if (!process.env.EXPORT_STATIC || !config.module) {
>       return config;
>     }
>     config.module.rules?.push({
>       test: /app\/api/,
>       loader: 'ignore-loader',
>     });
>     return config;
>   },
> ```
> 
> ```
> Error: Page "/api/auth/[...nextauth]" is missing "generateStaticParams()" so it cannot be used with "output: export" config.
> `
> ```

I got the same error.

**Comment by daveycodez** [comment]
I guess the only real way to fix this is to write a script that copies these files out of your project, does a build, then copies them back. ROFL! C'mon.... Just ignore API routes on exports obviously!

**Reply by icyJoseph** [reply]
Nah, just use custom page extensions, https://github.com/vercel/next.js/discussions/51891#discussioncomment-6297178

**Reply by daveycodez** [reply]
Oh that's pretty sick actually, the documentation doesn't describe it exactly this way but I'm bookmarking that

**Comment by Zulqarnain-tkxel** [comment]
same issue my middleware of routing not working with out : "export in next config does anyone has solution for this ?

**Reply by icyJoseph** [reply]
Explain again. You use Middleware on dev mode, but when you ship, you use export output, and the Middleware gets in the way?

**Reply by Zulqarnain-tkxel** [reply]
I have used middleware in my nextjs 14.2 where I am checking that if token is there or not . If there then its giving me access to private routes.
But issue lies when I have export:output in my nexjts config my Middleware is not triggering. 
when I comment out my line output:"export" then its working.. its in localhost not on prod I havenot pushed to prod and checked.
my middleware is in src/middlware. 
I can do with useEffect but I want to follow nextjs approach with middleware

My nextj config
/** @type {import('next').NextConfig} */
const nextConfig = {
  output: "export",
  // distDir: "build", // Replace 'build' with your desired folder name
  reactStrictMode: true,

};

my middlware 

import { NextResponse } from "next/server";

export const privateRoutes = ["/brand-shop", "/campaigns", "/order-history", "/framme-collection", "/manage-shop", "/profile"];

export function middleware(req) {
  // Get cookies from the request
  const cookies = req.cookies;
  const token = cookies.get("token")?.value || null;

  if (privateRoutes.includes(req.nextUrl.pathname) && !token) {
    // If it's a private route and no token is found, redirect to login
    return NextResponse.redirect(new URL("/", req.url));
  }

  // If it's a public route or a token is present, continue
  return NextResponse.next();
}

**Reply by daveycodez** [reply]
Output export doesn't support middleware
https://nextjs.org/docs/app/building-your-application/deploying/static-exports#unsupported-features

**Reply by icyJoseph** [reply]
Middleware is not supported in export mode, https://nextjs.org/docs/app/building-your-application/deploying/static-exports#unsupported-features

**Reply by artola** [reply]
> Explain again. You use Middleware on dev mode, but when you ship, you use export output, and the Middleware gets in the way?

I use middleware on development because I need to add `headers` before a `redirect` (this one does does not support headers, only destination). Then when building with `output="export"` the `middleware.ts` is still there and a warning is printed.

> Statically exporting a Next.js application via `next export` disables API routes and middleware.

It would be nice:
- to disable from the config, `middleware: false`, to indicate the intention.
- to be able to pass headers in the redirects, apart from destination, like a prop or like a function that allows to evaluate the request and return a redirect with/without new headers. IN such a way, I would not need a middleware for such a simple case.

**Reply by icyJoseph** [reply]
You can do this:

- https://github.com/vercel/next.js/discussions/51891#discussioncomment-6297178

Call it `middleware.api.ts` or js, and it'll work in dev mode, but be carved out, as the author wanted,  during production build, given that you follow my instructions there.

I think that should do the trick anyway, only ever tried with API routes, but Middleware is also affected by Page extensions, so why not. It is worth the try.

**Reply by artola** [reply]
@icyJoseph I will try right now. Thanks.

FYI it does not play so nice. If I use the extension "banana", there are problems with the loaders. Then I would need to do something like @daveycodez comment below, but it is just a hack on top of another hack. `middleware: false` would clearly state that even if the file exists, it is not mean to be loaded.

**Reply by daveycodez** [reply]
I just use this:
```ts
    output: "export",
    pageExtensions: ["jsx", "tsx"],
```
    
Since all my API Routes and middleware are "ts" they get ignored

**Comment by zizzkata** [comment]
There should be a feature that let's you exclude paths. This is so basic, why is this not here by default.
Simply using webpack doesn't get you passed 'Collecting page data' check during build. At least disable api routes during static export.
